packages/backend/src/models/ProjectModel/ProjectModel.ts

```diff
@@ -442,4 +444,31 @@ export class ProjectModel {
             await func();
         });
     }
+
+    async getWarehouseFromCache(
+        projectUuid: string,
+    ): Promise<WarehouseCatalog> {
+        const warehouses = await this.database(CachedWarehouseTableName)
+            .select(['warehouse'])
+            .where('project_uuid', projectUuid)
+            .limit(1);
+        if (warehouses.length > 0) return warehouses[0].warehouse;
+        return {};
+    }
+
+    async saveWarehouseToCache(
+        projectUuid: string,
+        warehouse: WarehouseCatalog,
+    ): Promise<void> {
+        this.database.transaction(async (trx) => {
+            await trx(CachedWarehouseTableName)
+                .where('project_uuid', projectUuid)
+                .delete();
+
+            const result = await trx(CachedWarehouseTableName).insert({
+                project_uuid: projectUuid,
+                warehouse: JSON.stringify(warehouse),
+            });
```

Comment:
Rather than deleting then inserting (which could create a race condition), you can also upsert by using `ON CONFLICT` in postgres:

https://github.com/lightdash/lightdash/blob/cd7b666ea7a1de02138223400c530c2cde147233/packages/backend/src/models/ProjectModel/ProjectModel.ts#L148-L156

I also noticed this for `saveProjectToCache`

Reply:
I like :+1: